### PR TITLE
Back end side of system config command

### PR DESCRIPTION
**Pull request by jluhrs**
Implementation of the Single System Configuration command, that allows the user to apply a step configuration to a system.
The command was connected to the UI.
The feedback in the UI is not yet implemented.

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 1]
```diff
@@ -408,6 +401,46 @@ class SeqexecEngine(httpClient: Client[IO], settings: Settings[IO], sm: SeqexecM
     }).flatMap(_.map(Function.tupled(runQueue)).getOrElse(executeEngine.unit))
   }
 
+  private def configSystemCheck(sid: Observation.Id, sys: Resource)(st: EngineState): Boolean = {
+    // Resources used by running sequences
+    val used = resourcesInUse(st)
+
+    // TODO: this is too much guessing. We should have proper tracking of systems' state.
+    val failedInstruments: Set[Resource] = st.sequences.values.toList.mapFilter(s =>
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: is this duplicated from `SeqexecEngine`?

File: modules/seqexec/web/server/src/main/scala/seqexec/web/server/http4s/SeqexecCommandRoutes.scala [hunk 2]
```diff
@@ -173,8 +171,7 @@ class SeqexecCommandRoutes(auth:       AuthenticationService,
         Ok(s"Stopped from queue $qid")
 
     case POST -> Root / "execute" / ObsIdVar(oid) / PosIntVar(step) / ResourceVar(resource) as _ =>
-      // Just simulate a delay
-      IO.sleep(2.seconds) *>
+      se.configSystem(inputQueue, oid, step, resource) *>
```
Comment (cquiroz) [inline modules/seqexec/web/server/src/main/scala/seqexec/web/server/http4s/SeqexecCommandRoutes.scala]: So we are not using the response. How is it sent to the client?
How could the client know that a response is correlated with this request?

File: modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala [hunk 3]
```diff
@@ -13,13 +13,19 @@ import seqexec.server.SeqexecServerArbitraries._
...
+import seqexec.engine.{Action, Actions, Result}
 
 /**
   * Tests SeqexecServer Lenses
   */
 final class SeqexecServerLensesSpec extends CatsSuite with ArbObservation {
 
-  implicit val steppEq: Eq[HeaderExtraData => engine.Step[IO]] = Eq.fromUniversalEquals
+  // I tried to go down the rabbit hole with the Eqs, but it is not worth it for what they are used.
+  implicit val streamEq: Eq[fs2.Stream[IO, Result]] = Eq.fromUniversalEquals
```
Comment (cquiroz) [inline modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala]: This doesn't look right. You shouldn't be able to compare `Stream[IO, N]` I think

File: modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala [hunk 4]
```diff
@@ -13,13 +13,19 @@ import seqexec.server.SeqexecServerArbitraries._
...
+  implicit val actStateEq: Eq[Action.State] = Eq.fromUniversalEquals
+  implicit val actionEq: Eq[Action[IO]] = Eq.by(x => (x.kind, x.gen, x.state))
```
Comment (cquiroz) [inline modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala]: Same here, i don't think you can do `Eq` with `IO` involved

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 5]
```diff
@@ -617,39 +667,59 @@ object SeqexecEngine extends SeqexecConfiguration {
...
+   */
+  private def resourcesReserved(failedSystems: Set[Resource], st: EngineState): Set[Resource] = {
+    def reserved(q: ExecutionQueue): Set[Resource] = q.queue.fproduct(st.sequences.get).collect{
+      case (_, Some(s)) if s.seq.status.isIdle &&
+        failedSystems.intersect(s.seqGen.resources).isEmpty => s.seqGen.resources
+    }.foldK
+
+    val runningQs = st.queues.values.filter(_.status(st).running)
+
+    runningQs.map(reserved).toList.foldK
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: `foldK`?

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 6]
```diff
@@ -461,11 +494,22 @@ class SeqexecEngine(httpClient: Client[IO], settings: Settings[IO], sm: SeqexecM
...
     val seq = st.toSequence
+
+    def resources(s: SequenceGen.StepGen): List[Resource] = s match {
+      case SequenceGen.PendingStepGen(_, _, resources, _) => resources.toList
+      case _                                              => List.empty
+    }
     def engineSteps(seq: Sequence[IO]): List[Step] = {
 
       // TODO: Calculate the whole status here and remove `Engine.Step.status`
       // This will be easier once the exact status labels in the UI are fixed.
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: is this comment still relevant?

**Comment by cquiroz** [comment]
Some general questions:

* will the result of the state be stored on the server? e.g. if another client arrives after a resurce run, will it see it?
* Has this been tested with the real hardware? Should it be?

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala (hunk 5 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: :sparkles: magic :sparkles:

**Comment by jluhrs** [comment]
Some general answers:

- Yes, the server state tracks when a configuration has started and when it completes or fails. Therefore, this state is visible to all the clients.
- It has not been tested with real hardware. It has being proved that the command triggers the 'configure' method of the simulation instances of TCC, GCAL and GMOS. There is no difference between triggering that an the real systems.

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala (hunk 5 above)
Comment (tpolecat) [reply modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: `List(a, b, c).combineAll` uses `Monoid` and gives you `a |+| b |+| c` whereas `.foldK` uses `MonoidK` and gives you `a <+> b <+> c`.

`MonoidK` is a universally-quantified monoid; i.e., `forAll a. Monoid[F[a]]`.  So a hand-wavy intuition is that the element type is unknown so you can't look at the elements. For example `Monoid[Option[Int]]` adds up any `Int`s it sees but `MonoidK[Option]` can't look at the elements so it uses `orElse`.

```scala
@ Option(1) |+| None |+| Option(2) 
res12: Option[Int] = Some(3)

@ Option(1) <+> None <+> Option(2) 
res13: Option[Int] = Some(1)
```

File: modules/seqexec/web/server/src/main/scala/seqexec/web/server/http4s/SeqexecCommandRoutes.scala (hunk 2 above)
Comment (jluhrs) [reply modules/seqexec/web/server/src/main/scala/seqexec/web/server/http4s/SeqexecCommandRoutes.scala]: I'll create a Jira task to add the `clientId`

File: modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala [hunk 7]
```diff
@@ -13,19 +12,18 @@ import seqexec.server.SeqexecServerArbitraries._
...
   */
 final class SeqexecServerLensesSpec extends CatsSuite with ArbObservation {
 
   // I tried to go down the rabbit hole with the Eqs, but it is not worth it for what they are used.
-  implicit val streamEq: Eq[fs2.Stream[IO, Result]] = Eq.fromUniversalEquals
   implicit val actStateEq: Eq[Action.State] = Eq.fromUniversalEquals
-  implicit val actionEq: Eq[Action[IO]] = Eq.by(x => (x.kind, x.gen, x.state))
-  implicit val steppEq: Eq[HeaderExtraData => List[Actions[IO]]] = Eq.fromUniversalEquals
-  implicit val stepActionsGenEq: Eq[StepActionsGen] = Eq.by(x => (x.pre))
+  implicit def actionEq[F[_]]: Eq[Action[F]] = Eq.by(x => (x.kind, x.state))
```
Comment (cquiroz) [inline modules/seqexec/server/src/test/scala/seqexec/server/SeqexecServerLensesSpec.scala]: Maybe these shouldn't be `implicit` and pass them eplicitly to the tests instead

**Pull request merged**